src/box/lua/config/configdata.lua

```diff
@@ -594,6 +594,36 @@ local function validate_failover(found, peers, failover, leader)
     end
 end
 
+-- Validate failover section.
+local function validate_failover_config(cconfig, failover_config)
+    if not failover_config or not failover_config.replicasets then
+        return
+    end
+    for replicaset_name, replicaset in pairs(failover_config.replicasets) do
+        local replicaset_config = nil
+        for _, group in pairs(cconfig.groups) do
+            replicaset_config = group.replicasets and
+                group.replicasets[replicaset_name] or
+                replicaset_config
+        end
+
+        if not replicaset_config then
+            error(('replicaset %s specified in the failover configuration '..
+                   'doesn\'t exist'):format(replicaset_name), 0)
+        else
+            -- Validate the priority section of the specific replicasets.
+            for instance_name, _ in pairs(replicaset.priority) do
+                if not replicaset_config.instances or
+                   not replicaset_config.instances[instance_name] then
+                    error(('instance %s from replicaset %s specified in the '..
+                           'failover configuration doesn\'t exist')
+                          :format(instance_name, replicaset_name), 0)
+                end
+            end
+        end
+    end
+end
```

Comment:
Nit: It feels a bit complicated for me. How about something like the following?

```diff
diff --git a/src/box/lua/config/configdata.lua b/src/box/lua/config/configdata.lua
index 5b4ec17538..91b60b1335 100644
--- a/src/box/lua/config/configdata.lua
+++ b/src/box/lua/config/configdata.lua
@@ -595,30 +595,32 @@ local function validate_failover(found, peers, failover, leader)
 end
 
 -- Validate failover section.
-local function validate_failover_config(cconfig, failover_config)
-    if not failover_config or not failover_config.replicasets then
+local function validate_failover_config(instances, failover_config)
+    if failover_config == nil or failover_config.replicasets == nil then
         return
     end
-    for replicaset_name, replicaset in pairs(failover_config.replicasets) do
-        local replicaset_config = nil
-        for _, group in pairs(cconfig.groups) do
-            replicaset_config = group.replicasets and
-                group.replicasets[replicaset_name] or
-                replicaset_config
-        end
 
-        if not replicaset_config then
+    local replicasets = {}
+    for _, def in pairs(instances) do
+        replicasets[def.replicaset_name] = true
+    end
+
+    for replicaset_name, replicaset in pairs(failover_config.replicasets) do
+        if replicasets[replicaset_name] == nil then
             error(('replicaset %s specified in the failover configuration '..
                    'doesn\'t exist'):format(replicaset_name), 0)
-        else
-            -- Validate the priority section of the specific replicasets.
-            for instance_name, _ in pairs(replicaset.priority) do
-                if not replicaset_config.instances or
-                   not replicaset_config.instances[instance_name] then
-                    error(('instance %s from replicaset %s specified in the '..
-                           'failover configuration doesn\'t exist')
-                          :format(instance_name, replicaset_name), 0)
-                end
+        end
+
+        -- Validate the priority section of the specific replicasets.
+        for instance_name, _ in pairs(replicaset.priority or {}) do
+            if instances[instance_name] == nil then
+                error(('instance %s from replicaset %s specified in the '..
+                       'failover configuration doesn\'t exist')
+                      :format(instance_name, replicaset_name), 0)
+            end
+
+            if instances[instance_name].replicaset_name ~= replicaset_name then
+                error('XXX', 0)
             end
         end
     end
@@ -834,7 +836,7 @@ local function new(iconfig, cconfig, instance_name)
     validate_failover(found, peers, failover, leader)
 
     local failover_config = instance_config:get(iconfig_def, 'failover')
-    validate_failover_config(cconfig, failover_config)
+    validate_failover_config(instances, failover_config)
 
     local bootstrap_strategy = instance_config:get(iconfig_def,
         'replication.bootstrap_strategy')
```

Reply:
Thank you. Applied the suggested patch. Looks simpler than the previous version. The only thing I've changed the `XXX` error is now an `assert(false)`

Reply:
> The only thing I've changed the `XXX` error is now an `assert(false)`

This is a possible validation error (a user placed an instance in the `failover` section into a wrong replicaset). We should report a user-friendly diagnostic (error) in the case. (I would also add a test case for it.)

Reply:
My bad, haven't though about it. Thank you for this note. Added the message and the test.